samples/net/sockets/http_server/src/server.c

```diff
@@ -0,0 +1,979 @@
+{
+
+	printf("IDLE_STATE\n");
+
+	parse_http2_frame(ctx_client->buffer, ctx_client->offset, &frame);
+	print_http2_frames(&frame);
+
+	ctx_client->streams[ctx_client->stream_count].stream_id =
+		find_headers_frame_stream_id(&frame);
+
+	printf("||stream id %d||\n",
+		ctx_client->streams[ctx_client->stream_count].stream_id);
+
+	if (!has_upgrade_header) {
+		ctx_client->streams[0].stream_id = 1;
+		unsigned char response_headers_frame[16];
+
+		ctx_client->streams[ctx_client->stream_count].stream_state =
+			OPEN_STATE;
+
+		generate_response_headers_frame(response_headers_frame,
+			ctx_client->streams[0].stream_id);
+
+		size_t ret = sendall(ctx_client->client_fd,
+			response_headers_frame, sizeof(response_headers_frame));
+		if (ret < 0) {
+			LOG_ERR("ERROR writing to socket");
+			return -errno;
+		}
+
+		ctx_client->stream_count = 1;
+	}
+
+	if (settings_end_headers_flag(ctx_client->buffer[4])) {
+		unsigned char response_headers_frame[16];
+
+		generate_response_headers_frame(response_headers_frame,
+			ctx_client->streams[ctx_client->stream_count]
+				.stream_id);
+
+		ctx_client->stream_count++;
+
+		size_t ret = sendall(ctx_client->client_fd,
+			response_headers_frame, sizeof(response_headers_frame));
+		if (ret < 0) {
+			LOG_ERR("ERROR writing to socket");
+			return -errno;
+		}
+	} else {
+		/* We expect Continuation frame */
+		ctx_client->server_state = HTTP2_FRAME_HEADER_STATE;
+		return 0;
+	}
+
+	unsigned int frame_size = determine_frame_size(ctx_client->buffer);
+
+	ctx_client->offset -= frame_size;
+	memmove(ctx_client->buffer, ctx_client->buffer + frame_size,
+		ctx_client->offset);
+
+	if (ctx_client->offset == 0) {
+		if (ctx_client->streams[ctx_client->stream_count]
+				.stream_state == IDLE_STATE) {
+			handle_http2_open_state(ctx_client);
+		} else {
+			ctx_client->server_state = HTTP2_FRAME_HEADER_STATE;
+		}
+	} else {
+		ctx_client->server_state =
+			determine_server_state(ctx_client->buffer);
+	}
+
+	return 0;
+}
+
+int handle_http2_open_state(struct http2_client_ctx *ctx_client)
+{
+	printf("OPEN_STATE\n");
+
+	parse_http2_frame(ctx_client->buffer, ctx_client->offset, &frame);
+	print_http2_frames(&frame);
+
+	size_t content_size = sizeof(content);
+
+	for (int i = 0; i < ctx_client->stream_count; i++) {
+		send_data(ctx_client->client_fd, content, content_size, 0x00,
+			0x01, ctx_client->streams[i].stream_id);
+	}
+
+	ctx_client->streams[0].stream_state = -1;
+
+	if (has_upgrade_header)
+		if (ctx_client->offset == 0) {
+			ctx_client->server_state = HTTP2_FRAME_HEADER_STATE;
+		} else {
+			ctx_client->server_state =
+				determine_server_state(ctx_client->buffer);
+		}
+	else
+		ctx_client->server_state = HTTP2_FRAME_GOAWAY_STATE;
+
+	has_upgrade_header = true;
+
+	return 0;
+}
+
+int handle_http2_frame_headers(struct http2_client_ctx *ctx_client)
+{
+	printf("HTTP2_FRAME_HEADERS\n");
+
+	ctx_client->streams[0].stream_state =
+		determine_stream_state(ctx_client->buffer);
```

Comment:
Two issues with this approach:
 * First - we shouldn't be really using a fixed `streams` array index, as this effectively breaks dealing with multiple streams. Having a stream counter doesn't work either, as this will fail if we have multiple concurrent streams.
 * Second - I don't think there's a need for `determine_stream_state()` function. Especially, that with current use it isn't really useful (note that we check frame type in there, but we already know what kind of frame (HEADERS) since we're in respective state handlers.

For current stream identification, we should be using stream ID instead. Therefore, I suggest to introduce two helper functions: 
* `struct http2_stream_ctx *find_http2_stream_context(struct http2_client_ctx *ctx_client, uint32_t stream_id)`
* `struct http2_stream_ctx *allocate_http2_stream_context(struct http2_client_ctx *ctx_client, uint32_t stream_id)`

First would loop over clients `struct http2_stream_ctx streams` and return the stream context with a matching ID, NULL if not found.
Second, would loop over clients `struct http2_stream_ctx streams` and look for an "empty" stream context (i. e. stream in IDLE state). If not found, return NULL.

Additionally, `struct http2_stream_ctx *current_stream` inside client context could be useful, to remember which stream the current frame belongs to.

Now how I'd use those functions. That's where the aforementioned state transition function can be handy. For example, we've received frame header which indicates that the frame type is HEADERS frame, we call `enter_http2_frame_headers_state()` :

* First, we use `find_http2_stream_context()` to find if the corresponding stream already exists
* If not (NULL returned) we use `allocate_http2_stream_context()` and try to allocate new stream (remember, that only HEADERS frame can do that, i. e. spawn new streams!).
* In case no empty stream can be found, reject the stream (or for now just close the connection for simplicity).
* Initialize any helper variables that may be needed to track frame parsing progress (`need_pad_len` etc,  https://discord.com/channels/@me/1105531307448074281/1134064198775750658 to refresh your memory).
* Otherwise, mark the stream as OPEN and enter the `HTTP2_FRAME_HEADERS_STATE`
* The list can be extended, if there are additional conditions stated in the RFC (or implied by the implementation).